### ISSUE TITLE
Filter leak when using WithSubscriber

### ISSUE THREAD

**Issue by jeromegn** [tokio-rs/tracing#902]
### Version

```
├── tracing v0.1.18 (*)
├── tracing-core v0.1.13 (*)
├── tracing-futures v0.2.4
│   └── tracing v0.1.18 (*)
├── tracing-log v0.1.1
│   └── tracing-core v0.1.13 (*)
├── tracing-subscriber v0.2.10
│   ├── tracing-core v0.1.13 (*)
│   ├── tracing-log v0.1.1 (*)
│   └── tracing-serde v0.1.1
│       └── tracing-core v0.1.13 (*)
```

### Platform

Linux 5.7.12-arch1-1

### Crates

tracing-subscriber (probably)

### Description

When using the `WithSubscriber` trait, the filter appears to leak to the global subscriber. At least when not setting any filter.

I'm using a global subscriber with an `EnvFilter`. If I use something like:

 ```rust
my_future.with_subscriber(tracing_subscriber::Registry::default())
```

then once my future completes, my global subscriber has "changed" (I think) and no longer filter events or spans according to the original setting. In my use case I was setting the level to "info" and I started getting events and spans with traces up to the "trace" level.

**Comment by hawkw** [comment]
I think I know what's up here. When a `Dispatch` is created, we call the `callsite::register_dispatch` function, which adds the `Dispatch` to the registry, and registers every callsite with it:
https://github.com/tokio-rs/tracing/blob/7f2428354fa0b6e107d6ac0998440475cd9913b5/tracing-core/src/dispatcher.rs#L409

This stores cached filter evaluations from all current subscribers in the per-callsite cache.

However, when a `Dispatch` is _dropped_, these cached filters remain in the callsite cache until the next time it is invalidated, either by creating a new `Dispatch` or by manually calling [`rebuild_interest_cache`](https://docs.rs/tracing-core/0.1.13/tracing_core/callsite/fn.rebuild_interest_cache.html). This results in the dropped dispatcher's filters persisting.

I think we ought to change this so that whenever the _last_ clone of a `Dispatch` is dropped, it re-invalidates the interest cache. I think the easiest way to do this might be changing the `Arc<dyn Subscriber + Send + Sync>` to a newtype of some kind, and having a `Drop` impl for that type that calls `rebuild_interest_cache`.

**Comment by hawkw** [comment]
@jeromegn the fix I described above should be fairly straightforward to implement, if you're interested in getting another `tracing-core` commit under your belt. :) Otherwise, I'm happy to do it.

**Comment by jeromegn** [comment]
I can give it a shot! Will probably ask for help on Discord though :)

**Comment by jeromegn** [comment]
Confirmed fixed by 2a5da8d
